lbrynet/core/Session.py

```diff
@@ -193,30 +196,56 @@ def _try_upnp(self):
 
         log.debug("In _try_upnp")
 
-        def get_free_port(upnp, port, protocol):
-            # returns an existing mapping if it exists
+        def get_mapped_port(upnp, port, protocol):
+            # returns an existing mapped port if it exists
             mapping = upnp.getspecificportmapping(port, protocol)
-            if not mapping:
-                return port
-            if upnp.lanaddr == mapping[0]:
-                return mapping
-            return get_free_port(upnp, port + 1, protocol)
-
-        def get_port_mapping(upnp, internal_port, protocol, description):
-            # try to map to the requested port, if there is already a mapping use the next external
-            # port available
+            if mapping is not None and upnp.lanaddr == mapping[0]:
+                return mapping[1]
+            return None
+
+        def add_port_mapping(upnp, port, auto_port, protocol, description):
+            log.debug('UPNP automatic port selection enabled: ' + str(auto_port))
+            try:
+                if auto_port is True:
+                    # addanyportmapping() returns None on failure
+                    ext_port = upnp.addanyportmapping(port, protocol, upnp.lanaddr, port,
+                                                      description, '')
+                else:
+                    # addportmapping() raises an exception on failure
+                    upnp.addportmapping(port, protocol, upnp.lanaddr, port, description, '')
+                    ext_port = port
+            except Exception as e:
+                log.warning('UPnP failed with error: %s', e.message)
+                return None
+
+            return ext_port
```

Comment:
A bit shorter and tidied up:
```
    def add_port_mapping(upnp, port, auto_port, protocol, description):
        log.debug('UPNP automatic port selection enabled: {}'.format(auto_port))
        try:
            if auto_port:
                return upnp.addanyportmapping(
                    port, protocol, upnp.lanaddr, port, description, ''
                )
            else:
                upnp.addportmapping(port, protocol, upnp.lanaddr, port, description, '')
                return port
        except Exception as e:
            log.warning('UPnP failed with error: %s', e.message)
```

Reply:
I normally prefer to keep the returns out, but if like this style, I'll follow that.
For the function arguments, I really prefer to see them aligned to the opening parenthesis, unless we have a codestyle suggesting not to

Reply:
If the function has finished and you have the value to return it's good to just return the value. Someone reading your code will quickly see that after the appropriate value is computed nothing else happens. If you only put returns at the end of methods it makes it seem like other stuff happens. Also, it require temporary variables and just more lines of code. More lines of code means more technical debt.

I don't know if there is an official lbry style on this particular point. You can leave that part as you had it until we figure out if we need a consistent style. Personally, I find that lining up arguments with parenthesis leads to longer lines and big gaps of empty space below the function and to the left of the parenthesis. If you put arguments on a new line and indent 4 spaces you can fit more arguments, more concisely.

Reply:
more lines of code are good if they increase readability. But I understand in this case I see there is no real gain.

Imho having the return at the end of the function with a specific variable makes you understand that, no matter what happens, we will return the content of that little box. How it is computed may be of interest or may not.

While having the returns into indented blocks forces you to dig in to understand what we are really returning. Well, I guess in python you'd need to do that anyway because there is not strict typing...so ext_port may contain any type of object..

I'll move the returns inside.

About the extra variable: I am used to compiled languages where you don't really care about it because it gets optimized out. Often in the linux kernel extra variables are used to avoid long conditions/expressions/lines and it works pretty well in my experience.
But I guess in python this does not happen as it is a scripted language with sequential parsing. I'll remove it.

Thanks for your feedback